Improve Password Policy

momo3404 opened the pull request:
Fixes weak password policy.

Changes proposed in this PR:

- Add password complexity requirements to ensure all new passwords, in addition to already requiring a minimum length of 8 characters, also have 1 uppercase letter, 1 lowercase letter, 1 digit, and 1 special character.
- Add live validation to ensure users understand requirements as they're typing their passwords.
- NOTE: Users with weak passwords will NOT be required to change their passwords after this change. This will only apply to account creation, password editing, and password resetting.

momo3404 committed:
Add password complexity validation and error messages

momo3404 committed:
Add password_checklist css and dmp-assistant stylesheet block

momo3404 committed:
Add password_checklist html

momo3404 committed:
Add password_checklist javascript

momo3404 committed:
Fix breaking tests

momo3404 committed:
Add password complexity and validation tests

momo3404 committed:
Rename password-requirements to password-checklist

momo3404 committed:
Refactor user_spec tests

momo3404 committed:
Add Changelog Entry

aaronskiba committed:
Add `const` to `passwordRequirements`

aaronskiba committed:
Wrap passwordChecklist.js in function / enable early return

- This change adds the early return of `passwordChecklist.js` when a page doesn't include any relevant elements (i.e. `if (passwordInputs.length === 0 || !passwordRequirements) return;`).
- To enable this early return, we are now assigning the prior code to the `passwordChecklist` function, and then calling it immediately.

aaronskiba committed:
Update passwordChecklist.js to reduce DOM queries

Add `ruleIconsMap` to map password rules to their corresponding `<i>` icons.
- Eliminates repeated DOM queries previously being performed in `updateRequirements()`

aaronskiba committed:
Simplify passwordChecklist.js to single password input

Replaced the `passwordsInputs` array with a single `passwordInput` variable. Pages never have more than one password field that uses `passwordField.js`, and this change should simplify the code a little bit.

aaronskiba committed:
Rename variables in `passwordChecklist.js`

- Renamed `passwordRequirements` to `passwordChecklist` to reflect elements actual name
- Renamed `passwordChecklist` function name to `initPasswordChecklist` to better distinguish it from the newly added `passwordChecklist` variable
- Renamed `updateRequirements` to `updateChecklistIcons` to clarify behaviour

aaronskiba commented on app/views/devise/passwords/edit.html.erb:
```diff
@@ -17,7 +17,8 @@
       </div>
       <div class="form-group">
         <%= f.label(:password_confirmation, _('Password confirmation'), class: 'control-label') %>
-        <%= f.password_field(:password_confirmation, class: 'form-control', "aria-required": true, autocomplete: "new-password") %>
```
It looks like the update should've been performed on the `f.password_field` above this one.

momo3404 committed:
Correct password validation in password reset

aaronskiba committed:
Add comments to `passwordChecklist.js`

aaronskiba committed:
Remove redundant explicit `password_field` IDs

- Rails automatically assigns `id="user_password"` to password fields, so
explicit `id: 'user_password'` attributes were unnecessary.

- Additionally, the manually defined `id: 'user_password'` in `app/views/shared/_create_account_form.html.erb` had no effect. Rather, `id="new_user_password"` was being generated due to `namespace: 'new'` being specified in the form.

aaronskiba committed:
Merge pull request #1162 from portagenetwork/aaron/passwordChecklistJS-updates

`passwordChecklist.js` Changes

aaronskiba reviewed (approved):
Big security improvement here. Nice work. 👍
